### PR TITLE
Show radio station name in settings

### PR DESCRIPTION
**Pull request by martinposchmann**
Here is a new branch containing only my cherry-picked commits.
This branch saves the radio station as json string into the settings and displays the name, country and bitrate as settings summary.
The stream url will be displayed as fallback, if the json string is not available from settings.

### CODE REVIEW

File: src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java [hunk 1]
```diff
@@ -239,6 +243,32 @@ protected Object onGetDefaultValue(TypedArray a, int index) {
     }
 
     @Override
+    public CharSequence getSummary () {
+        RadioStation station = getPersistedRadioStation();
+        if (station != null && station.name != null && !station.name.isEmpty()) {
+            String summary = station.name;
+            if (station.countryCode != null || station.bitrate > 0) {
+                summary += " (";
+                boolean hasCountry = false;
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java]: boolean hasCountry = (station.countryCode != null);

File: src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java [hunk 2]
```diff
@@ -380,5 +430,25 @@ private String getDisplayedRadioStationText(RadioStation station, boolean displa
         }
     }
 
+    private void persistRadioStation(RadioStation station) {
+        //save stream as separate field
+        persistString(station.stream);
+
+        //save complete station as json string
+        Settings settings = new Settings(getContext());
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java]: You should use getSharedPreferences() to access the prefs in a way which is independent of the nightdream package.
Derive a key from the key of the preference.

SharedPreferences prefs = getSharedPreferences();
String key = String.format("%s_json", getKey());

File: src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java [hunk 3]
```diff
@@ -380,5 +430,25 @@ private String getDisplayedRadioStationText(RadioStation station, boolean displa
...
+        settings.setRadioStation(station);
+    }
+
+    private RadioStation getPersistedRadioStation() {
+
+        Settings settings = new Settings(getContext());
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java]: See above.

File: src/com/firebirdberlin/radiostreamapi/models/RadioStation.java [hunk 4]
```diff
@@ -13,4 +16,43 @@ public String toString() {
...
+        RadioStation station = new RadioStation();
+        station.id = jsonStation.getLong(JSON_ID);
+        station.name = jsonStation.getString(JSON_NAME);
+        station.countryCode = jsonStation.getString(JSON_COUNTRY);
+        station.stream = jsonStation.getString(JSON_STREAM);
+        String bitRateString = jsonStation.getString(JSON_BITRATE);
+        if (bitRateString != null && !bitRateString.isEmpty()) {
+            try {
+                station.bitrate = jsonStation.getLong(JSON_BITRATE);
+            } catch (Throwable t) {
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/radiostreamapi/models/RadioStation.java]: Why don't you catch JSONException ?

File: src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java [hunk 5]
```diff
@@ -239,6 +243,32 @@ protected Object onGetDefaultValue(TypedArray a, int index) {
...
+                boolean hasCountry = false;
+                if (station.countryCode != null) {
+                    summary += station.countryCode;
+                    hasCountry = true;
+                }
+                if (station.bitrate > 0) {
+                    if (hasCountry) {
+                        summary += ", ";
+                    }
+                    summary += String.valueOf(station.bitrate) + " kbps";
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java]: I like String.format() more ...

File: src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java (hunk 1 above)
Comment (martinposchmann) [inline src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java]: Yes, thats way better, I've updated the code.

File: src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java (hunk 5 above)
Comment (martinposchmann) [inline src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java]: Uses String.format() now

File: src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java [hunk 6]
```diff
@@ -38,6 +38,10 @@
 import com.firebirdberlin.radiostreamapi.models.Country;
 import com.firebirdberlin.radiostreamapi.models.RadioStation;
 
+import org.json.JSONException;
+
+import static com.firebirdberlin.nightdream.Settings.PREFS_KEY;
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java]: Not needed. See below.

File: src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java [hunk 7]
```diff
@@ -435,20 +439,36 @@ private void persistRadioStation(RadioStation station) {
         persistString(station.stream);
 
         //save complete station as json string
-        Settings settings = new Settings(getContext());
-        settings.setRadioStation(station);
+        try {
+            String json = station.toJson();
+            SharedPreferences preferences = mContext.getSharedPreferences(PREFS_KEY, 0);
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java]: This one is much simpler: 

SharedPreferences preferences = getSharedPreferences();

File: src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java [hunk 8]
```diff
@@ -435,20 +439,36 @@ private void persistRadioStation(RadioStation station) {
...
+            SharedPreferences.Editor prefEditor = preferences.edit();
+            prefEditor.putString("radioStation", json);
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java]: Please build the key the following way:

String key = String.format("%s_json", getKey());

File: src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java [hunk 9]
```diff
@@ -435,20 +439,36 @@ private void persistRadioStation(RadioStation station) {
...
+            prefEditor.commit();
+        } catch (Throwable t) {
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java]: This try block refers to station.toJson() only. You should use the JSONException.

File: src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java [hunk 10]
```diff
@@ -435,20 +439,36 @@ private void persistRadioStation(RadioStation station) {
...
+
     }
 
     private RadioStation getPersistedRadioStation() {
 
-        Settings settings = new Settings(getContext());
         try {
-            RadioStation station = settings.getRadioStation();
-            return station;
+            SharedPreferences preferences = mContext.getSharedPreferences(PREFS_KEY, 0);
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java]: See above ...

File: src/com/firebirdberlin/radiostreamapi/models/RadioStation.java (hunk 4 above)
Comment (martinposchmann) [inline src/com/firebirdberlin/radiostreamapi/models/RadioStation.java]: I changed it

File: src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java (hunk 2 above)
Comment (martinposchmann) [inline src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java]: Done.

**Comment by martinposchmann** [comment]
Applied all of your requested changes. I squashed all commits into a single one.

File: src/com/firebirdberlin/nightdream/Settings.java [hunk 11]
```diff
@@ -18,14 +18,19 @@
 import android.Manifest;
 import android.os.Build;
 import android.support.v4.content.ContextCompat;
+import android.util.Log;
 
 import static android.text.format.DateFormat.getBestDateTimePattern;
 
 import com.firebirdberlin.nightdream.models.BatteryValue;
 import com.firebirdberlin.nightdream.models.WeatherEntry;
+import com.firebirdberlin.radiostreamapi.models.RadioStation;
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/Settings.java]: No longer needed. This is true for the other new imports, too.

File: src/com/firebirdberlin/radiostreamapi/DirbleApi.java [hunk 12]
```diff
@@ -99,13 +99,17 @@
...
                         RadioStation station = new RadioStation();
                         station.id = jsonStation.getLong("id");
                         station.name = jsonStation.getString("name");
                         station.countryCode = jsonStation.getString("country");
                         station.stream = streamObj.getString("stream");
-                        station.bitrate = streamObj.getLong("bitrate");
+                        if (streamObj.get("bitrate") != null) {
+                            try {
+                                station.bitrate = streamObj.getLong("bitrate");
+                            } catch (Throwable t) {
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/radiostreamapi/DirbleApi.java]: JSONException

File: src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java [hunk 13]
```diff
@@ -120,7 +124,8 @@ public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
             @Override
             public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                 RadioStation station = stations.get(position);
-                persistString(station.stream);
+                //persistString(station.stream);
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java]: To be removed.

File: src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java [hunk 14]
```diff
@@ -380,5 +429,42 @@ private String getDisplayedRadioStationText(RadioStation station, boolean displa
...
+        //save stream as separate field
+        persistString(station.stream);
+
+        //save complete station as json string
+        try {
+            String json = station.toJson();
+            SharedPreferences prefs = getSharedPreferences();
+            SharedPreferences.Editor prefEditor = prefs.edit();
+            prefEditor.putString(jsonKey(), json);
+            prefEditor.commit();
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/radiostreamapi/RadioStreamPreference.java]: JSONException ?

**Comment by martinposchmann** [comment]
All change requests are now fixed in the code.

**Pull request merged**